extensions/clearinghouse/src/main/java/clearinghouse/IdsClearingHouseServiceImpl.java

```diff
@@ -0,0 +1,88 @@
+import org.eclipse.edc.connector.contract.spi.types.agreement.ContractAgreement;
+import org.eclipse.edc.protocol.ids.service.ConnectorServiceSettings;
+import org.eclipse.edc.spi.EdcException;
+import org.eclipse.edc.spi.event.Event;
+import org.eclipse.edc.spi.event.EventSubscriber;
+import org.eclipse.edc.spi.event.contractnegotiation.ContractNegotiationConfirmed;
+import org.eclipse.edc.spi.message.RemoteMessageDispatcherRegistry;
+import org.eclipse.edc.spi.monitor.Monitor;
+import org.eclipse.edc.spi.system.Hostname;
+import sender.message.LogMessage;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+
+public class IdsClearingHouseServiceImpl implements IdsClearingHouseService, EventSubscriber {
+
+    private static final String CONTEXT_CLEARINGHOUSE = "ClearingHouse";
+
+    private final RemoteMessageDispatcherRegistry dispatcherRegistry;
+    private final ConnectorServiceSettings connectorServiceSettings;
+    private final Hostname hostname;
+    private final URL clearingHouseLogUrl;
+    private final ContractNegotiationStore contractNegotiationStore;
+    private final Monitor monitor;
+
+    public IdsClearingHouseServiceImpl(
+            RemoteMessageDispatcherRegistry dispatcherRegistry,
+            ConnectorServiceSettings connectorServiceSettings,
+            Hostname hostname,
+            URL clearingHouseLogUrl,
+            ContractNegotiationStore contractNegotiationStore,
+            Monitor monitor) {
+        this.dispatcherRegistry = dispatcherRegistry;
+        this.connectorServiceSettings = connectorServiceSettings;
+        this.hostname = hostname;
+        this.clearingHouseLogUrl = clearingHouseLogUrl;
+        this.contractNegotiationStore = contractNegotiationStore;
+        this.monitor = monitor;
+    }
+
+    @Override
+    public void logContractAgreement(ContractAgreement contractAgreement, URL clearingHouseLogUrl) {
+        monitor.info("Logging contract agreement to ClearingHouse");
+        try {
+            var connectorBaseUrl = new URI(String.format("http://%s/", hostname.get()));
+            var logMessage = new LogMessage(clearingHouseLogUrl, connectorBaseUrl, contractAgreement);
+
+            dispatcherRegistry.send(Object.class, logMessage, () -> CONTEXT_CLEARINGHOUSE);
+        } catch (URISyntaxException e) {
+            throw new EdcException("Could not create connectorBaseUrl. Hostname can be set using:" +
+                    " edc.hostname", e);
+        }
+    }
+
+    @Override
+    public void on(Event<?> event) {
+        if (event instanceof ContractNegotiationConfirmed contractNegotiationConfirmed) {
+            //ContractNegotiationConfirmed -> Log to ClearingHouse
+            handleContractNegotiationConfirmed(contractNegotiationConfirmed);
+        }
+    }
+
+    private void handleContractNegotiationConfirmed(ContractNegotiationConfirmed contractNegotiationConfirmed) {
```

Comment:
Refactor method to private ContractAgreement resolveContractAgreement(ContractNegotiationConfirmed contractNegotiationConfirmed)

remove `logContractAgreement(contractAgreement, clearingHouseLogUrl);` call and call the method in the `public void on(Event<?> event)` method instead